### PR TITLE
Don't try to read unknown formats in builtin reader plugin

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
Currently, the builtin reader plugin _always_ returns a function, even if it's bound to fail.  This means that if it's not in the last position, it will always try, and you'll see an error message.

This PR gates the reader function on a [set of extensions](https://github.com/napari/napari/pull/3145/files#diff-920df4d91a5a168fa33f90c37220bb6304de548aebb6bb4fce86cd4aa6b7e10fR18-R19) (the imageio extensions plus a couple more).  Would appreciate some double checking on the extension choices.

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

**Review by alisterburt — APPROVED** [review]
I'm not sure about the extensions but this looks good to me, could maybe add a test to check that what used to raise an error doesn't with this change?

**Review by goanpeca — APPROVED** [review]

**Comment by Carreau** [comment]
This fails some of the Zarr tests, as with zarr you can open a subfolder, and the tests are trying to open `temp.zarr/1d/`.

I'm not too sure how to handle that, as both opening a directory as a zarr file, and enforcing .zarr make sens.

**Comment by tlambert03** [comment]
> I'm not too sure how to handle that, as both opening a directory as a zarr file, and enforcing .zarr make sens.

perhaps we can let all directories continue to go through... that would still "silence" most of the unnecessary errors i'm seeing when dragging a single file

File: napari/plugins/_builtins.py [hunk 1]
```diff
@@ -57,7 +58,10 @@ def napari_get_reader(path: Union[str, List[str]]) -> ReaderFunction:
     """
     if isinstance(path, str) and path.endswith('.csv'):
         return csv_reader_function
-    return image_reader_to_layerdata_reader(magic_imread)
+
+    _p = [path] if isinstance(path, str) else path
+    if all(str(x).lower().endswith(tuple(READER_EXTENSIONS)) for x in _p):
+        return image_reader_to_layerdata_reader(magic_imread)
```
Comment (Carreau) [inline napari/plugins/_builtins.py:67]: We _could_ special case zarr to make sure that
`any(p.endswith('.zarr') for p in Path(stuff).absolute().parts)`.

**Comment by Carreau** [comment]
> perhaps we can let all directories continue to go through... that would still "silence" most of the unnecessary errors i'm seeing when dragging a single file

Yeah, letting dirs go through seem like a good idea as well. Simpler than my proposal.

**Review by sofroniewn — APPROVED** [review]
LGTM! This had 3 test fails, which is a little more then I think we're normally seeing right now, so I'm rerunning them. If it's less I'll merge, but we might need to keep an eye on this

**Pull request merged**